apps/website/app/utils/supabase/dbUtils.ts

```diff
@@ -0,0 +1,482 @@
+
+export type BaseEntityResult = {
+  error: string | null;
+  details?: string; // For detailed error messages, e.g., from Supabase
+  status: number; // HTTP status code to suggest
+};
+
+const processSupabaseError = (
+  error: PostgrestError,
+  tableName: string,
+): BaseEntityResult => {
+  console.error(`Error inserting new ${tableName}:`, error);
+  if (error.code === "23505") {
+    // Handle race condition: unique constraint violation (PostgreSQL error code '23505')
+    console.warn(
+      `Unique constraint violation on ${tableName} insert .\nError is ${error.message}, ${error.hint}.`,
+    );
+    const dup_key_data = unique_key_re.exec(error.hint);
+    if (dup_key_data !== null && dup_key_data.length > 1) {
+      const uniqueOn = dup_key_data[1]!.split(",").map((x) => x.trim());
+      if (uniqueOn.length === 0) {
+        const idx_data = unique_index_re.exec(error.message);
+        return {
+          error:
+            idx_data === null
+              ? "Could not identify the keys or index of the unique constraint"
+              : `Could not identify the keys of the unique key index ${idx_data[1]}`,
+          details: error.message,
+          status: 400,
+        };
+      }
+    }
+    return {
+      error: `Unique constraint violation on ${tableName} insert, and re-fetch failed to find the entity.`,
+      details: error.message,
+      status: 409, // Conflict, and couldn't resolve by re-fetching
+    };
+  }
+  if (error.code === "23503") {
+    // Handle foreign key constraint violations (PostgreSQL error code '23503')
+    const fkey_data = foreign_key_re.exec(error.hint);
+    const constraint_data = foreign_constraint_re.exec(error.message);
+    console.warn(
+      `Foreign violation on ${tableName}, constraint ${constraint_data ? constraint_data[2] : "unknown"}, keys ${fkey_data ? fkey_data[1] : "unknown"}`,
+    );
+    return {
+      error:
+        fkey_data === null
+          ? constraint_data === null
+            ? "Could not identify the missing foreign key"
+            : `Foreign key constraint ${constraint_data[2]} violated`
+          : `Foreign key ${fkey_data[1]} is missing value ${fkey_data[2]}`,
+      details: error.message,
+      status: 400,
+    };
+  }
+
+  return {
+    error: `Database error ${error.code} while upserting in ${tableName}.`,
+    details: error.message,
+    status: 500,
+  };
+};
+
+export type GetOrCreateEntityResult<T> = BaseEntityResult & {
+  entity: T | null;
+  created: boolean;
+};
+
+/**
+ * Generic function to get an entity or create it if it doesn't exist.
+ * Handles common race conditions for unique constraint violations.
+ *
+ * @param supabase Supabase client instance.
+ * @param tableName The name of the table.
+ * @param insertData Data to upsert
+ * @param uniqueOn The expected uniqueOn key.
+ * @returns Promise<GetOrCreateEntityResult<T>>
+ */
+export const getOrCreateEntity = async <
```

Comment:
`get` here is misleading.  Isn't it technically `createOrUpdateEntity` or `upsertEntity`?

This follows for all of it's callers as well. Let's unify them where possible. If the caller is also just upserting, let's have the name match it's function.  Currently, some are `create`, `internal`, or other.